### PR TITLE
 create_test_timeline: always put@initdb_lsn the minimum required keys

### PR DESCRIPTION
**Pull request by problame**
See the added comment on `create_empty_timeline`.

The various test cases now need to set a valid `Lsn` instead of `Lsn(0)`.

Rough context: https://github.com/neondatabase/neon/pull/4364#discussion_r1221995691

### CODE REVIEW

File: pageserver/src/tenant.rs [hunk 1]
```diff
@@ -4353,6 +4377,29 @@ mod tests {
...
+    async fn test_empty_test_timeline_is_usable() -> anyhow::Result<()> {
+        let (tenant, ctx) = TenantHarness::create("test_empty_test_timeline_is_usable")?
+            .load()
+            .await;
+        let tline =
+            tenant.create_test_timeline(TIMELINE_ID, Lsn(0x20), DEFAULT_PG_VERSION, &ctx)?;
+
+        // Make sure the timeline has the minimum set of required keys for operation.
+        // The only operation you can do on an empty timeline is to write new data.
+        // Repartition is the only code on the write path that requires other keys to be present.
```
Comment (hlinnaka) [inline pageserver/src/tenant.rs:4391]: This feels a little misleading: many operations in walingest.rs read the existing keys.

File: pageserver/src/tenant.rs [hunk 2]
```diff
@@ -1316,7 +1328,19 @@ impl Tenant {
...
+        // Setup minimum keys required for the timeline to be usable.
+        let mut modification = uninit_tl
+            .raw_timeline()
+            .expect("we just created it")
+            .begin_modification(initdb_lsn);
+        modification.init_empty().context("init_empty")?;
+        modification
+            .commit()
+            .context("commit init_empty modification")?;
+
```
Comment (hlinnaka) [inline pageserver/src/tenant.rs:1341]: In `pgdatadir_mapping::create_test_timeline`, there's also a call to `init_empty` which is now redundant with this.

Comment (problame) [reply pageserver/src/tenant.rs:1341]: Yep, learned that, fixing it.

**Comment by problame** [comment]
There is some race condition in the tests. When running each one individually they don't fail. When running them in parallel, they fail.

I think they fail due to ephemeral file page writeback failing.

Debugging...

**Comment by problame** [comment]
(I since figured out the race condition and the code is ready for re-review)

File: pageserver/src/pgdatadir_mapping.rs [hunk 3]
```diff
@@ -1601,9 +1601,6 @@ pub fn create_test_timeline(
     ctx: &RequestContext,
 ) -> anyhow::Result<std::sync::Arc<Timeline>> {
     let tline = tenant.create_test_timeline(timeline_id, Lsn(8), pg_version, ctx)?;
-    let mut m = tline.begin_modification(Lsn(8));
-    m.init_empty()?;
-    m.commit()?;
     Ok(tline)
 }
```
Comment (hlinnaka) [inline pageserver/src/pgdatadir_mapping.rs:1605]: Does it make sense to have this function anymore? Let's just inline it to the callers.

**Review by hlinnaka — APPROVED** [review]

File: pageserver/src/tenant.rs [hunk 4]
```diff
@@ -1268,6 +1268,18 @@ impl Tenant {
...
+    /// and the timeline will fail to load at a restart.
+    ///
+    /// That's why we add an uninit mark file, and wrap it together witht the Timeline
+    /// in-memory object into UninitializedTimeline.
+    /// Once the caller is done setting up the timeline, they should call
+    /// `UninitializedTimeline::initialize_with_lock` to remove the uninit mark.
+    ///
+    /// For tests, use `DatadirModification::init_empty_test_timeline` + `commit` to setup the
+    /// minimum amount of keys required to get a writable timeline.
+    /// (Without it, `put` might fail due to `repartition` failing.)
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant.rs:1282]: Ah shute, that function moved to the next PR on top of this one: https://github.com/neondatabase/neon/pull/4456

Will fixup.

**Pull request merged**